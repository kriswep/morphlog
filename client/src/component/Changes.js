import React from 'react';
import { gql } from 'apollo-boost';
import { Query, Mutation } from 'react-apollo';
import styled from 'styled-components';
import { Button, Comment } from 'semantic-ui-react';

import Input from './Input';

const Author = styled.div`
  display: inline-block;
`;

const CHANGES_QUERY = gql`
  query changes($projectId: ID!) {
    changes(projectId: $projectId) {
      id
      createdAt
      text
      author {
        id
        name
      }
    }
  }
`;

const ADD_CHANGE_MUTATION = gql`
  mutation addChange($projectId: ID!, $txt: String!) {
    addChange(projectId: $projectId, text: $txt) {
      id
      text
      createdAt
      author {
        id
        name
      }
      project {
        id
      }
    }
  }
`;

const initialState = {
  text: '',
};

class Project extends React.Component {
  state = initialState;

  dispatch = e => {
    if (e.target.name && e.target.name) {
      const newState = {};
      newState[e.target.name] = e.target.value;
      this.setState(newState);
    }
  };

  addChange = mutate => {
    mutate({
      variables: {
        projectId: this.props.projectId,
        txt: this.state.text,
      },
      update: (proxy, { data: { addChange } }) => {
        // Read the data from our cache for this query.
        const data = proxy.readQuery({
          query: CHANGES_QUERY,
          variables: {
            projectId: addChange.project.id,
          },
        });

        // Add our todo from the mutation to the end.
        const newChanges = [addChange, ...data.changes];

        // Write our data back to the cache.
        proxy.writeQuery({
          query: CHANGES_QUERY,
          variables: {
            projectId: addChange.project.id,
          },
          data: { changes: newChanges },
        });
      },
      refetchQueries: [
        {
          query: CHANGES_QUERY,
          variables: {
            projectId: this.props.projectId,
          },
        },
      ],
    })
      .then(({ data }) => {
        console.log('got data', data);
      })
      .then(() => {
        this.setState(initialState);
      })
      .catch(error => {
        console.log('there was an error sending the query', error);
        this.setState(initialState);
      });
  };

  render() {
    return (
      <div>
        <Mutation mutation={ADD_CHANGE_MUTATION}>
          {(mutate, { error }) => (
            <div data-test="newChange">
              <Input
                textarea
                name="text"
                label="Change"
                value={this.state.text}
                type="text"
                placeholder="describe your change"
                onChange={this.dispatch}
              />
              <Button
                color="teal"
                onClick={e => {
                  e.preventDefault();
                  this.addChange(mutate);
                }}
              >
                Add
              </Button>
              {error && `Error!: ${error}`}
            </div>
          )}
        </Mutation>
        <Query
          query={CHANGES_QUERY}
          variables={{
            projectId:
              this.props.projectId || this.props.match.params.projectId,
          }}
        >
          {({ loading, error, data }) => {
            if (loading) return null;
            if (error) return `Error!: ${error}`;

            const { changes } = data;
            if (!changes) return null;
            return (
              <Comment.Group data-test="change">
                {changes.map(change => (
                  <Comment key={change.id}>
                    <Comment.Content>
                      <Comment.Author as={Author}>
                        {change.author.name}
                      </Comment.Author>
                      <Comment.Metadata>
                        <div>{new Date(change.createdAt).toLocaleString()}</div>
                      </Comment.Metadata>
                      <Comment.Text>{change.text}</Comment.Text>
                      <Comment.Actions>
                        <Comment.Action>Maybe Change</Comment.Action>
                      </Comment.Actions>
                    </Comment.Content>
                  </Comment>
                ))}
              </Comment.Group>
            );
          }}
        </Query>
      </div>
    );
  }
}

export default Project;
