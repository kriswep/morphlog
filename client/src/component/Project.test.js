/* globals test expect jest */
// TODO checkout https://github.com/react-cosmos/react-cosmos#react-apollo-graphql

import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import ProjectWithApollo, { Project } from './Project';
import Changes from './Changes';

import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';

jest.mock('./Changes', () => {
  return () => <div>MockedChange</div>;
});

const typeDefs = `
  type Project {
    id: ID
    name: String
  }
  type Query {
    project(id: ID): Project
  }
`;

const mocks = {
  Query: () => ({
    project: () => ({
      id: 'cjcp94wxp025801100npb28yg',
      name: 'MyProject',
      createdAt: '2018-01-21T20:45:06.000Z',
      updatedAt: '2018-01-21T20:45:06.000Z',
    }),
  }),

  // Mutation: () => ...
};

const schema = makeExecutableSchema({ typeDefs });
addMockFunctionsToSchema({
  schema,
  mocks,
});

const apolloCache = new InMemoryCache(window.__APOLLO_STATE__);

const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({ schema }),
});

test('Projects renders correctly', async () => {
  const wrapper = mount(
    <ApolloProvider client={client}>
      <ProjectWithApollo projectId="cjcp94wxp025801100npb28yg" />
    </ApolloProvider>,
  );
  await new Promise(res => window.setTimeout(res, 100));
  wrapper.setProps({ projectId: 'bar' });
  expect(toJson(wrapper.find(Project))).toMatchSnapshot();
});
