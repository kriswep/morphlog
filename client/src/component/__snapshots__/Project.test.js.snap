// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects renders correctly 1`] = `
<Project
  projectId="cjcp94wxp025801100npb28yg"
  projectQuery={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "project": Object {
        "__typename": "Project",
        "id": "cjcp94wxp025801100npb28yg",
        "name": "MyProject",
        Symbol(id): "Project:cjcp94wxp025801100npb28yg",
      },
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "cjcp94wxp025801100npb28yg",
      },
    }
  }
>
  <styled.section>
    <section
      className="sc-bdVaJa fmkqYI"
    >
      <div>
        <h2>
          MyProject
        </h2>
        <Component
          projectId="cjcp94wxp025801100npb28yg"
        >
          <div>
            MockedChange
          </div>
        </Component>
      </div>
    </section>
  </styled.section>
</Project>
`;
